Upgrade to Yarn 4, remove support for Node 16

renchap opened the pull request:
Yarn 1 is no longer maintained and suffers from many bugs (including https://github.com/eslint/eslint/discussions/17215#discussioncomment-6609097 which prevents multiple package upgrades)

This also bumps the minimum version to Node 18, as Node 16 is EOL and no longer supported by Yarn

I followed the official upgrade guide, most of the changes are switching to the new CLI syntax.

Our setup should now properly support `corepack`, so installing the package manager is as simple as `corepack prepare` (and maybe `corepack enable` before, if it has never been ran on this machine and is not down when installing node). It will automatically pick the correct yarn version.

We could want to switch to `pnpm` or plain `npm`, as those are 2 viable alternatives nowadays, but Webpacker requires a `yarn` command in many places, so this is the easiest way to get a modern JS package manager. This can be changed later is needed.

vmstan commented:
Running Yarn 3 compiled setup on vmst.io now.

ClearlyClaire commented:
Haven't tested it yet, as my main dev setup uses distribution-provided NodeJS and… `corepack` is not part of the distribution yet :|

This seems fine, but I'm not very comfortable with shipping minified yarn right in the repo. That seems to be how yarn 3 works, though.

vmstan commented:
I've yet to run into issues with builds.

renchap committed:
Upgrade to Yarn 4

renchap commented:
Upgraded the PR to use the new Yarn 4. It no longer requires having the `yarn` executable in-repo, instead relying on `corepack` to have the correct `yarn` version installed (this is the new recommended Yarn setup).

The changes are minimal, compared to switching to plain `npm` or `pnpm`, which getting better performance and useful features like `yarn patch`.

renchap commented:
This is now based on #27518 to have the custom Github actions to setup the JS env, as `setup-node` does not support Corepack yet, and it is needed for Yarn 4.

renchap committed:
Add a custom yarn version check

renchap committed:
Fix Dockerfile

nschonni commented on package.json:
```diff
@@ -2,8 +2,11 @@
   "license": "AGPL-3.0-or-later",
   "engines": {
-    "node": ">=16"
+    "node": ">=18"
```
Seems like it's at 20 in the NVM and Dockerfile
```suggestion
    "node": ">=20"
```

vmstan replied:
But 18 is still supported

nschonni replied:
It is, but only 20 is being tested on and shipped with the images

renchap replied:
We do not have real tests for the JS part running on Node, ie the streaming server, so this is not really an issue.

We could try to test the Webpack build with both Node 18 and Node 20, but I do not feel this would be very useful, as this is not running our code.

renchap committed:
Apply local fix from https://github.com/lodash/babel-plugin-lodash/pull/261

Babel changed their interface and `babel-plugin-lodash` is no longer maintained, but the patch is trivial.

renchap committed:
Use `peerDependenciesMeta` to remove warning about unmatched peer dependencies versions

This makes those 3 packages optional peer deps, which removes the warning, as recommended in Yarn's doc

renchap commented:
Rebased, and added:
- the needed configuration so the peer-dependency warnings are no longer displayed (fixes #27611). It transforms the peer deps for those packages into optional ones
- a local patch for `babel-plugin-lodash` to remove fix the deprecation warning on assets compilation. Patch is from https://github.com/lodash/babel-plugin-lodash/pull/261 (fixes #27550)

renchap committed:
Copy `.yarn` to container image so patches are there when installing

renchap committed:
Replace the custom `compression-webpack-plugin` with a local patch

ClearlyClaire reviewed (approved):
This sounds like a simple enough change, and I could easily install yarn 4 with and without corepack.

However, a few caveats:
- `yarn set version 4` does not work from `yarn` 1, but `yarn set version 4.0.1` and `yarn set version berry` both do
- at some point while playing with yarn versions, `yarn` 4.0.1 got installed to `.yarn/releases` and `.yarnrc.yml` changed to explicitly reference it; I am not sure what caused this, but this has potential to cause issues for admins
